### DISCUSSION TITLE
Sometimes user needs to click on the canvas to start the emulation (Autoplay policy?)

### DISCUSSION THREAD

**Discussion by gitGalu** [arianrhodsandlot/nostalgist#41] [Q&A] [answered]
This might be similar to someone's question in the issues section.

Sometimes, after starting emulation via await Nostalgist.launch(...), the emulation launches but does not start until the user clicks or taps on the emulator canvas. The screen doesn't necessarily remain black - sometimes it's a different color (perhaps the first frame of the emulated image?). I assume this isn't a bug but rather a browser security measure against autoplay, as this situation usually occurs when there’s a longer delay between clicking the element that starts the emulation and loading all files (more than a few seconds), though not every time.

Is it possible to detect this situation (e.g., in the onLaunch callback) to display a notification? This would help avoid always informing the user that they need to click, as in most cases, the emulation starts automatically.

**Comment by arianrhodsandlot** [comment, marked as answer]
> I assume this isn't a bug but rather a browser security measure against autoplay

You are right :) We cannot bypass it.

> Is it possible to detect this situation

Unfortunately we can't detect whether is this happening. And different browsers on different platforms have different limitations about the time between user actions and autoplay of media.

Basically, there are two ways to deal with this problem.

+ Separate the fetching process of ROM/core files from the launching call, so when the user clicks something, we can guarantee that the related files have already been loaded.
  Example:
  ```js
  let rom
  async function loadRom() {
    const response = await fetch('/path/to/rom')
    rom = await response.blob()
  }
  loadRom()

  document.querySelector('.launch-button').addEventListener('click', () => {
    if (!rom) {
      alert('rom is not ready yet')
    }

    Nostalgist.launch({
      core: 'anycore',
      rom: {
        fileName: 'filename.rom',
        fileContent: rom
      }
    })
  }
  ```

+ Measure the launching time, if it takes longer than a maximum number, show a button to ask the user to click the button. That maximum number varies from different browsers and devices, so maybe we should use different maximum numbers for them.

  ```js
  // These numbers are for demonstration purposes only, they are not accurate.
  let max = 500
  if (navigator.userAgent.includes('iPhone')) {
    max = 300
  } else if (navigator.userAgent.includes('Android')) {
    max = 400
  }

  document.querySelector('.launch-button').addEventListener('click', () => {
    const start = Date.now()
    Nostalgist.launch({
      core: 'anycore',
      rom: 'path/to/rom',
      async beforeLaunch() {
        const end = Date.now()
        const time = end - start
        if (time > max) {
          // the files that will be used by the emulator are all loaded at this phase
          // so we can show another button to ask user to click it
          await new Promise((resolve, reject) => {
            // the `modal` variable may come from a UI library or something we implement by ourselves
            modal.show({
              text: 'It takes too long to load the game, so you have to click the confirm button to continue!'
              onConfirm() {
                resolve()
              }
            })
          })
        }
      }
    })
  }
  ```

**Reply by gitGalu** [reply]
Thanks!
Loading some WASM files can also contribute to slow loading times on slower connections, so I'll fetch them in advance and modify the resolveCoreWasm() as well:

                resolveCoreWasm(file) {
                    if (wasmArrayBuffer) {
                        return wasmArrayBuffer;
                    }
                    return `./libretro/${core}_libretro.wasm`
                },
